File: apps/ios/Mishran/UI/Addresses/AddressesViewModel.swift
// AddressesViewModel.swift — Task 48.2 (Mishran Mobile Apps v1).
// Account → Delivery addresses state (parity with Android's 10.2
// AddressesViewModel): list/set-default/create/delete over
// AddressRepository, mirroring every mutation back into the SwiftData
// AddressEntity cache so checkout's @Query-backed picker stays in sync.
import Foundation
import Observation
import SwiftData

@MainActor
@Observable
final class AddressesViewModel {
    private let repository: AddressRepository
    private let context: ModelContext

    private(set) var addresses: [AddressDTO] = []
    private(set) var isLoading = false
    /// One-line status for the screen ("Could not set the default. Try
    /// again.") — mirrors Android's snackbar copy.
    var message: String?

    init(repository: AddressRepository, context: ModelContext) {
        self.repository = repository
        self.context = context
    }

    /// Refetch the server list, then swap the local cache to match.
    func refresh() async {
        isLoading = true
        defer { isLoading = false }
        addresses = await repository.list()
        AddressEntity.replaceAll(with: addresses, in: context)
    }

    /// Create via the form; refreshes the list or reports the failure.
    func create(_ input: AddressInputDTO) async -> Bool {
        guard await repository.create(input: input) != nil else {
            message = "Could not save the address. Try again."
            return false
        }
        message = nil
        await refresh()
        return true
    }

    /// PATCH the address with isDefault=true (the server demotes the
    /// previous default; refresh picks the change up).
    func setDefault(_ address: AddressDTO) async {
        guard let id = address.id, address.isDefault != true else { return }
        let updated = await repository.update(id: id, input: AddressInputDTO(address: address, isDefault: true))
        if updated != nil {
            message = nil
            await refresh()
        } else {
            message = "Could not set the default. Try again."
        }
    }

    /// Swipe-to-delete; a failure keeps the row and reports why.
    func delete(_ address: AddressDTO) async {
        guard let id = address.id else { return }
        if await repository.delete(id: id) {
            message = nil
            await refresh()
        } else {
            message = "Could not delete the address. Try again."
        }
    }
}

File: apps/android/app/src/main/java/com/mishran/app/ui/addresses/AddressesViewModel.kt
// apps/android/app/src/main/java/com/mishran/app/ui/addresses/AddressesViewModel.kt
//
// Account → Delivery addresses: list, add, set-default, and delete over the
// addresses routes (GET/POST/PATCH/DELETE). Errors surface as a one-line
// message instead of a dead screen.
package com.mishran.app.ui.addresses

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mishran.api.models.Address
import com.mishran.api.models.AddressInput
import com.mishran.app.data.repository.AddressRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class AddressesUiState(
    val loading: Boolean = true,
    val addresses: List<Address> = emptyList(),
    val message: String? = null,
)

@HiltViewModel
class AddressesViewModel @Inject constructor(
    private val repository: AddressRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(AddressesUiState())
    val state: StateFlow<AddressesUiState> = _state.asStateFlow()

    init {
        refresh()
    }

    fun refresh() {
        viewModelScope.launch {
            val addresses = repository.listAddresses()
            _state.update { it.copy(loading = false, addresses = addresses) }
        }
    }

    /** Create via the form; refreshes the list or reports the failure. */
    fun addAddress(input: AddressInput) {
        viewModelScope.launch {
            val created = repository.createAddress(input)
            if (created == null) {
                _state.update { it.copy(message = "Could not save the address. Try again.") }
            } else {
                _state.update { it.copy(message = null) }
                refresh()
            }
        }
    }

    /** PATCH the address with isDefault=true (server demotes the previous). */
    fun setDefault(address: Address) {
        viewModelScope.launch {
            val updated = repository.updateAddress(
                address.id.orEmpty(),
                address.toInput(isDefault = true),
            )
            if (updated == null) {
                _state.update { it.copy(message = "Could not set the default. Try again.") }
            } else {
                _state.update { it.copy(message = null) }
                refresh()
            }
        }
    }

    /** DELETE the address (owner-scoped); refreshes the list or reports the failure. */
    fun deleteAddress(address: Address) {
        viewModelScope.launch {
            val deleted = repository.deleteAddress(address.id.orEmpty())
            if (!deleted) {
                _state.update { it.copy(message = "Could not delete the address. Try again.") }
            } else {
                _state.update { it.copy(message = null) }
                refresh()
            }
        }
    }

    fun clearMessage() {
        _state.update { it.copy(message = null) }
    }
}

/** Address model → writable input; only the default flag changes on update. */
internal fun Address.toInput(isDefault: Boolean? = this.isDefault): AddressInput = AddressInput(
    line1 = line1.orEmpty(),
    line2 = line2,
    city = city.orEmpty(),
    state = state.orEmpty(),
    pincode = pincode.orEmpty(),
    tag = tag?.let { modelTag ->
        AddressInput.Tag.entries.firstOrNull { it.value == modelTag.value }
    },
    isDefault = isDefault,
)
